corehq/apps/es/client.py

```diff
@@ -435,27 +437,35 @@ def reindex(
             "conflicts": "proceed"
         }
 
-    # Should be removed after ES 5-6 migration
-        if copy_doc_ids and source == const.HQ_USERS_INDEX_NAME:
-            # Remove password from form index
+        if copy_doc_ids or purge_ids:
             reindex_body["script"] = {
                 "lang": "painless",
-                "source": """
-                ctx._source.remove('password');
-                if (!ctx._source.containsKey('doc_id')) {
-                    ctx._source['doc_id'] = ctx._id;
-                }
-                """
-            }
-        elif copy_doc_ids:
-            reindex_body["script"] = {
-                "lang": "painless",
-                "source": """
-                if (!ctx._source.containsKey('doc_id')) {
-                    ctx._source['doc_id'] = ctx._id;
-                }
-                """
+                "source": ""
             }
+            script_parts = []
+
+            if purge_ids:
+                script_parts.append("""
+                    if (ctx._source.containsKey('_id')) {
+                        ctx._source.remove('_id');
+                    }
+                """)
+
+            if source == const.HQ_USERS_INDEX_NAME:
+                # Remove password field from users index
+                script_parts.append("""
+                    ctx._source.remove('password');
+                """)
+
+            if copy_doc_ids:
+                # Add doc_id field to the documents
+                script_parts.append("""
+                    if (!ctx._source.containsKey('doc_id')) {
+                        ctx._source['doc_id'] = ctx._id;
```

Comment:
Yes. It will work if `purge_ids` is true. 

We are removing `_id` from `ctx._source` which is different from `ctx._id`. `ctx._id` is a read only property. 

`ctx._source.remove` is basically a groovy style syntax to remove an object from a map. So it is just removing the attribute from the object.